# 05 — Repetition System Part 1

## For Loops

### Anchor

A `for` loop is a conveyor belt.

Python takes one item at a time from a collection and applies the same action to each item.

Structure:

for item in collection:
    action

Meaning:

For each item inside this collection, do this action.

In [2]:
temperatures = [36.5, 37.2, 38.1]

for temperature in temperatures:
    print(temperature)

36.5
37.2
38.1


temperatures is the collection.
temperature is the temporary name for each item passing through.
for ... in ... means: take one item at a time and apply the action.

for temperature in temperatures:

means:

Take each item inside `temperatures`, call it `temperature` temporarily, and run the indented action.

In [5]:
names = ["Ana", "Luis", "Maya"]

for name in names:
    print(f"Hello {name}")

Hello Ana
Hello Luis
Hello Maya


name is the temporary name for each item passing through the collection names.

The f-string helps because it inserts the current value of name into the text.

In [6]:
temperatures = [36.5, 37.2, 38.1]

total = 0

for temperature in temperatures:
    total = total + temperature

print(total)

111.80000000000001


start: total = 0

cycle 1: temperature = 36.5
total = 0 + 36.5 → 36.5

cycle 2: temperature = 37.2
total = 36.5 + 37.2 → 73.7

cycle 3: temperature = 38.1
total = 73.7 + 38.1 → 111.8

total is an accumulator.
It remembers the growing result across loop cycles.

It does not add zero each time forever.
Zero is just the starting point.

In [11]:
temperatures = [36.5, 37.2, 38.1]

count = 0

for temperature in temperatures:
    count = count + 1

print(count)

3


start: count = 0

cycle 1: temperature = 36.5
count = 0 + 1 → 1

cycle 2: temperature = 37.2
count = 1 + 1 → 2

cycle 3: temperature = 38.1
count = 2 + 1 → 3

So count does not care what the temperature value is.
It only cares that one more item passed through the loop.

### Accumulator vs Counter

An accumulator adds the values themselves.

Example:
total = total + temperature

A counter counts how many loop cycles happened.

Example:
count = count + 1

Accumulator = how much in total  
Counter = how many items/cycles

In [12]:
temperatures = [36.5, 37.2, 38.1]

total = 0
count = 0

for temperature in temperatures:
    total = total + temperature
    count = count + 1

average = total / count

print(average)

37.26666666666667


The loop collects information over time.
After the loop, we can compute a final result from the accumulated state.

This is an important programming architecture pattern:

initialize state
↓
update state each cycle
↓
compute final result after loop

In [1]:
temperatures = [36.5, 37.2, 38.1, 39.0, 36.8]

fever_total = 0
fever_count = 0

for temperature in temperatures:
    if temperature >= 38:
        fever_total = fever_total + temperature
        fever_count = fever_count + 1

print(fever_total)
print(fever_count)

77.1
2


This combines for repetition, if decision, and memory updates through accumulation/counting.

In [2]:
temperatures = [36.5, 37.2, 38.1, 39.0, 36.8]

normal_count = 0
fever_count = 0
fever_total = 0

for temperature in temperatures:
    if temperature >= 38:
        fever_count = fever_count + 1
        fever_total = fever_total + temperature
    else:
        normal_count = normal_count + 1

fever_average = fever_total / fever_count

print(f"Normal readings: {normal_count}")
print(f"Fever readings: {fever_count}")
print(f"Average fever temperature: {fever_average}")

Normal readings: 3
Fever readings: 2
Average fever temperature: 38.55


This is the same core pattern, but now you added the fallback path

So now the loop classifies every item into one of two groups

A loop can scan a collection, classify each item, update memory, and produce a summary report.
This is the basic skeleton of data analysis.

# Session Summary — Repetition System Part 1

## Core anchor

A `for` loop is a conveyor belt.

Python takes one item at a time from a collection and applies the same action.

## Main concepts

- A collection contains multiple items.
- A loop variable is a temporary name for the current item.
- The loop runs one cycle per item.
- An accumulator stores a growing total.
- A counter stores how many cycles/items were counted.
- A loop can contain an `if` gate to filter items.

## Core pattern

initialize memory
↓
loop through collection
↓
update memory each cycle
↓
use final memory after loop

## Diagnostic question

Am I:
1. doing something to every item,
2. accumulating values,
3. counting cycles,
4. or filtering which items count?